content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md

```diff
@@ -0,0 +1,57 @@
+---
+title: Differences Between the EVM Emulator and Ethereum
+description:
+---
+::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
+Given that the EVM Bytecode Emulator runs on top of EraVM, the same limitations and differences detailed in
+the [Differences from Ethereum page](../../differences/evm-instructions) are inherited.
+::
+
+While the EVM Emulator is designed to execute EVM bytecode as closely as possible to Ethereum, there are several differences due
+to the underlying **EraVM execution model**.
+These differences affect gas accounting, opcode support, system behavior, and contract deployment.
+
+This page outlines the key differences between the **EVM Emulator on ZKsync** and **Ethereum (Cancun)**.
+
+## General Differences
+
+- **Gas accounting uses EraVM gas units (ergs) instead of EVM gas.**
+- **`JUMPDEST` analysis is simplified.** The emulator does not check whether `JUMPDEST` appears inside a `PUSH` instruction.
+- **EVM contract creation does not charge additional gas for initcode size** (`2 gas per 32-byte chunk` from
+  [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860) is omitted for EOAs and EraVM contract deployers but applied for EVM-to-EVM deployments).
+- **Call stack depth limit is not explicitly enforced.** Instead, it is naturally constrained by the `63/64` gas forwarding rule.
+- **EVM access lists (EIP-2930) are not supported.**
+- **Warm/cold storage logic differs.**
+  - Only contracts accessed within the EVM execution context are considered warm.
+  - Access from EraVM does **not** affect EVM warm/cold storage tracking.
+- **Contract storage is not cleared on deployment.** Unlike Ethereum, storage in newly created contracts is not explicitly reset.
+- **EVM gas refund mechanisms are not implemented.**
+- **Intrinsic gas costs differ from Ethereum.**
+- **Contract deployment failures increment nonces.** This matches Ethereum behavior but differs from EraVM,
+  where failed deployments do not affect the nonce.
+- **Nonce storage uses `u128` instead of `u64`.**
+
+## Gas Limit Behavior
+
+- The **`GASLIMIT` opcode returns EraVM batch limits in ergs**, not the Ethereum gas limit.
+- **Gas costs differ across opcodes.** Since EVM execution is emulated, each EVM instruction
+  requires multiple EraVM instructions, increasing overall execution costs.
+
+## Unsupported opcodes
+
+Some opcodes like CALLCODE, SELFDESTRUCT, BLOBHASH and BLOBBASEFEE are not supported
+on the EVM bytecode emulator as they're not supported on EraVM.
+
+Check [the list of unsupported opcodes](../../differences/evm-instructions)
+
+## Precompiles
+
+Some precompiles like `RIPEMD-160` and `blake2f` are not supported on the EVM bytecode
+emulator as they're not supported on EraVM.
+
+Check [the list of available precompiles](../../differences/pre-compiles)
```

Comment:
We need to add Identity precompile to the list of available precompiles

Reply:
added